src/com/unbounce/treajure/concurrent/mvar.clj

```diff
@@ -0,0 +1,82 @@
+(ns com.unbounce.treajure.concurrent.mvar
+  "Haskell's MVar abstraction in Clojure.
+
+  Check: http://chimera.labs.oreilly.com/books/1230000000929/ch07.html for more
+  info around MVars."
+  (:import [java.util.concurrent Semaphore]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Lightweight Performance mutable tuple implementation in Clojure
+;; (Look ma! no atoms!)
+
+(defprotocol IMCons
+  (mcar [p])
+  (mcdr [p])
+  (set-mcar! [p val])
+  (set-mcdr! [p val]))
+
+(deftype MCons [^{:volatile-mutable true} car
+                ^{:volatile-mutable true} cdr]
```

Comment:
@artgon @ddossot after trying to implement this abstraction with BlockingQueue, I figured out that this is _almost_ an MVar, but not quite the same behavior on the `read-mvar`

> peek()
> Retrieves, but does not remove, the head of this queue, or returns null if this queue is empty.

This is different from MVar, where it explicitly blocks if you try to read from it and it is empty. I think this implementation is not to bad, the semaphores are used consistently. The only thing that could be better in terms of understanding the code is the MCons type, this can be done with an atom of a tuple, but we would pay the runtime costs of atom's atomicity guarantees when they are not needed for the implementation.